component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -110,293 +110,477 @@ public AuthenticatorFlowStatus process(HttpServletRequest request,
+            if (log.isDebugEnabled()) {
+                log.debug("Default authentication endpoint context is used");
+            }
+        }
+        return errorPage;
+    }
+
+    /**
+     * Redirect to an error page.
+     *
+     * @param response    the HttpServletResponse
+     * @param queryParams the queryParams
+     * @throws AuthenticationFailedException
+     */
+    private void redirectToErrorPage(HttpServletResponse response, AuthenticationContext context, String queryParams,
+                                     String retryParam)
+            throws AuthenticationFailedException {
+        // that Enable the SMS OTP in user's Profile. Cannot proceed further without SMS OTP authentication.
+        try {
+            String errorPage = getErrorPage(context);
+            response.sendRedirect(errorPage + "?" + queryParams
+                    + SMSOTPConstants.AUTHENTICATORS + getName() + retryParam);
+        } catch (IOException e) {
+            throw new AuthenticationFailedException("Exception occurred while redirecting to errorPage. ", e);
+        }
+    }
+
+    /**
+     * In SMSOTP optional case proceed with first step only.It can be basic or federated.
+     *
+     * @param authenticatedUser the name of authenticatedUser
+     * @param context           the AuthenticationContext
+     */
+    private void processFirstStepOnly(AuthenticatedUser authenticatedUser, AuthenticationContext context) {
+        //the authentication flow happens with basic authentication.
+        StepConfig stepConfig = context.getSequenceConfig().getStepMap().get(context.getCurrentStep() - 1);
+        if (stepConfig.getAuthenticatedAutenticator().getApplicationAuthenticator() instanceof
+                LocalApplicationAuthenticator) {
+            FederatedAuthenticatorUtil.updateLocalAuthenticatedUserInStepConfig(context, authenticatedUser);
+            context.setProperty(SMSOTPConstants.AUTHENTICATION, SMSOTPConstants.BASIC);
+        } else {
+            FederatedAuthenticatorUtil.updateAuthenticatedUserInStepConfig(context, authenticatedUser);
+            context.setProperty(SMSOTPConstants.AUTHENTICATION, SMSOTPConstants.FEDERETOR);
+        }
+    }
+
+    /**
+     * Update mobile number when user forgets to update the mobile number in user's profile.
+     *
+     * @param context      the AuthenticationContext
+     * @param request      the HttpServletRequest
+     * @param username     the Username
+     * @param tenantDomain the TenantDomain
+     * @throws SMSOTPException
+     * @throws AuthenticationFailedException
+     */
+    private void updateMobileNumberForUsername(AuthenticationContext context, HttpServletRequest request,
+                                               String username, String tenantDomain)
+            throws SMSOTPException, AuthenticationFailedException {
+        if (username != null && !context.isRetrying()) {
```

Comment:
In getMobileNumber() method, first we have checked the mobile number whether it is already exist or not. If it is empty only, we proceed with the redirection to mobile number request page and update the mobile claim